若某研究者想探討某特定行業員工的年所得是否受性別、工作年資及教育年數所影響，於是隨機蒐集此行業員工的年所得、性別、工作年資及教育年數如附檔：(年所得單位：萬元)

請以性別、教育年數及工作年數為特徵變數，建立迴歸分析模式預測此行業員工年所得，並說明分析結果。

In [10]:
import pandas as pd
import numpy as np
data=pd.read_excel('salary2.xlsx')
data.head()

教育年數  工作年數  年所得  性別
0     6     7   38   1
1    12    14   53   0
2     9     5   40   1
3    16     8   50   0
4    18     6   55   1

In [11]:
import statsmodels.api as sm
X=data[['教育年數','工作年數','性別']]
y=data['年所得']
model=sm.OLS(y,X).fit()
print(model.summary())

                                 OLS Regression Results                                
Dep. Variable:                    年所得   R-squared (uncentered):                   0.988
Model:                            OLS   Adj. R-squared (uncentered):              0.988
Method:                 Least Squares   F-statistic:                              2450.
Date:                Thu, 29 Sep 2022   Prob (F-statistic):                    2.21e-85
Time:                        07:56:40   Log-Likelihood:                         -282.89
No. Observations:                  92   AIC:                                      571.8
Df Residuals:                      89   BIC:                                      579.3
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

R-squared和Adj. R-squared為0.988，很接近1，代表解釋能力好

教育年數、工作年數、性別的P值為0、0、0.05，小於0.05，代表有顯著性，拿來做預測是好的

In [12]:
X=data[['教育年數','工作年數','性別']]
y=data['年所得']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=333)
from sklearn.linear_model import LinearRegression
lm=LinearRegression()
lm.fit(X_train,y_train)
print(lm.intercept_) #截距
cdf=pd.DataFrame(lm.coef_,X_train.columns,columns=['Coef'])
cdf

24.62473505630998


Coef
教育年數  1.345679
工作年數  1.030251
性別   -0.491471

In [13]:
from sklearn.metrics import mean_squared_error
from math import sqrt
RMSE1=sqrt(mean_squared_error(y_train,lm.predict(X_train)))
RMSE2=sqrt(mean_squared_error(y_test,lm.predict(X_test)))
print('train_RMSE:',RMSE1)
print('test_RMSE:',RMSE2)

train_RMSE: 2.116533703745302
test_RMSE: 2.996310603707073


RMSE越高越不好，代表誤差越高

訓練<測試，所以測試時的誤差比較高

In [14]:
R2_train=lm.score(X_train,y_train)
R2_test=lm.score(X_test,y_test)
print('train_RSquare:',R2_train)
print('test_RSquare:',R2_test)

train_RSquare: 0.873712706503377
test_RSquare: 0.7482237891197683


R_Square越高越好，代表解釋能力越好

訓練>測試，所以訓練時的解釋能力比較好